nltk/translate/chrf_score.py

```diff
@@ -75,18 +84,31 @@ def sentence_chrf(reference, hypothesis, min_len=1, max_len=6, beta=3.0):
     :type max_len: int
     :param beta: the parameter to assign more importance to recall over precision
     :type beta: float
+    :param ignore_whitespace: ignore whitespace characters in scoring
+    :type ignore_whitespace: bool
     :return: the sentence level CHRF score.
     :rtype: float
     """
     return corpus_chrf([reference], [hypothesis], min_len, max_len, beta=beta)
 
 
-def corpus_chrf(list_of_references, hypotheses, min_len=1, max_len=6, beta=3.0):
+def _preprocess(sent, ignore_whitespace):
+    if type(sent) != str:
+        # turn list of tokens into a string
+        sent = ' '.join(sent)
+
+    if ignore_whitespace:
+        sent = re.sub(r'\s+', '', sent)
+    return sent
```

Comment:
Hmmm, this is the awkward part of chrf where the string is taken as an input.

Reply:
I think we can avoid doing this. By duck-typing 2 separate function and expose the `corpus_chrf_char` and `corpus_chrf_token` functions. And underspecify what the default `corpus_chrf` takes. 

Explicit is better than implicit?

Reply:
I don't like this type checking either, I was just trying to maintain the same interface. Two separate function names make sense, although I'm not sure I would be able to guess the difference between `*_char` and `*_token` from the name. (I assume the first would take sentences as strings and tokenize them whereas the other would take already tokenized lists?)

Reply:
This is really interesting problem of API design. I thought of doing


```
def _corpus_chrf(...):
    # generic function that is agnostic to input. 
    # When given list(str) -> character chrf
    # When given list(list(str)) -> token chrf

def corpus_chrf_char(...):
    corpus_chrf(...)

def corpus_chrf_token(...):
    corpus_chrf(...)

def corpus_chrf_token_char(...):
    corpus_chrf_token(...)
    corpus_chrf_char(...)

def sent_chrf_char(...):
    corpus_chr_char(...)

def sent_chrf_token(...):
    corpus_chr_char(...)
```

But that creates many level of spaghetti code =(

Reply:
def _corpus_chrf(...):
    # generic function that is agnostic to input. 
    # When given list(str) -> character chrf
    # When given list(list(str)) -> token chrf

def corpus_chrf(references, hypotheses, min_char_n, max_char_n, min_tok_n, max_tok_n, ...):
    _corpus_chrf(references, hypotheses, min_char_n, max_char_n, ...)
    references_char, hypotheses_char = preprocess(references), preprocess(hypotheses)
    _corpus_chrf(references_char, hypotheses_char, min_tok_n, max_tok_n, ...)

def sent_chrf(min_char_n, max_char_n, min_tok_n, max_tok_n, ...):
    corpus_char(min_char_n, max_char_n, min_tok_n, max_tok_n, ...):

Reply:
I was trying to reimplement the 2nd design but there are a couple averaging decisions invovled for `corpus_chrf` , looking at Maja's ChrF++ code.

 - average between token and char scores 
 - average between ngrams orders
 - average between sentences